### PR TITLE
httplog: add new logger

### PR DESCRIPTION
**Pull request by Choraden**
This patch adds new http logging modes

### CODE REVIEW

**Comment by mmatczuk** [comment]
This is quite inflexible. Also, how do we add separate access log to here?

**Comment by mmatczuk** [comment]
I do not like how the packages (forwarder and httplog) work together - it's not very flexible.
Also, it's not integrated at all with the logging logic we have in middleware - we have diverged.
We need a single flexible solution.

I'm not sure if it's ok to mix http mw with martian modifiers but maybe it's ok since both deal with http anyways.
With that in mind one component can have two applications as http.Handler wrapper and as a modifier.

Going back here I think it would be OK to add [middleware.Logger()](https://github.com/saucelabs/forwarder/blob/61b72918f99dc42dafa99c25c78aa0277519a718/middleware/logger.go#L19) function as a parameter in HTTPServerConfig.
This gives much flexibility and removes the coupling.

httplog can than expose functions that match middleware.Logger signature.
The function selection can be moved to bind module.
It's very easy to write a binding to anything including a function pointer with anyflag library.

**Comment by Choraden** [comment]
I appreciate your suggestions. The code is definitely more flexible and better integrated with middleware components. 

The drawback that's on my mind is that middleware.Logger does not allow to log whole http.Responses (with headers and body). But maybe we don't need to log them? 

My other concern is how to properly count written bytes without delegator. 

PTAL @mmatczuk

File: httplog/logger.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	ErrOnlyLogMode LoggerMode = "error"
+)
+
+func (m LoggerMode) String() string {
+	return string(m)
+}
+
+// Logger is a modifier that logs requests and responses.
+type Logger struct {
+	log func(line string)
```
Comment (mmatczuk) [inline httplog/logger.go:36]: How about using `Writer io.Writer`?

Then ignore the buffering here! Just use a buffered writer and add separate option for access logs.

Comment (mmatczuk) [reply httplog/logger.go:36]: Writer is no go due to to concurrency.

File: httplog/logger.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+// logging the body. Log function defaults to martian.Infof.
+func NewLogger() *Logger {
+	return &Logger{
+		log: func(line string) {
+			log.Infof(line)
+		},
+	}
+}
+
+func (l *Logger) Parse(val string) (middleware.Logger, error) {
```
Comment (mmatczuk) [inline httplog/logger.go:49]: There is a problem with configuration you want to do too much at once. A logger needs a log sink and mode.
This should be read from the config and only then you can properly initialize it...
SetLogFunc should be removed.

Defaulting to martian log is wrong, for instance now HttpServer uses matianlog...

File: middleware/logger.go [hunk 3]
```diff
@@ -28,9 +31,39 @@ func (l Logger) Wrap(h http.Handler) http.Handler {
...
 			Request:  r,
+			Response: nil,
 			Status:   d.Status(),
 			Written:  d.Written(),
 			Duration: time.Since(start),
 		})
 	})
 }
+
+const startTimeKey = "forwarder-middleware-start-time"
```
Comment (mmatczuk) [inline middleware/logger.go:42]: How about `start-time`.

File: httplog/logger.go (hunk 1 above)
Comment (mmatczuk) [inline httplog/logger.go:36]: How about we used the logging function signature `func(format string, args ...interface{})`

File: middleware/logger.go [hunk 4]
```diff
@@ -28,9 +31,39 @@ func (l Logger) Wrap(h http.Handler) http.Handler {
 		h.ServeHTTP(w, r)
 		l(LogEntry{
 			Request:  r,
+			Response: nil,
```
Comment (mmatczuk) [inline middleware/logger.go:34]: Not needed

**Comment by Choraden** [comment]
I decided to keep LogMode in ServerConfig again and create logger in-place. It was unnatural to initialise logFunc in bind. You would rather take things from config than put them there.

Creating Logger now takes logFunc `func(format string, args ...interface{})` and mode which is more flexible and configurable. 

We could think of creating separate named stdlogger for this, but I don't reckon it's necessary.  

PTAL @mmatczuk

**Pull request merged**